apis/python/src/tiledbsoma/options/_soma_tiledb_context.py

```diff
@@ -259,17 +261,19 @@ def replace(
                 new_config = self._internal_tiledb_config()
                 new_config.update(tiledb_config)
                 tiledb_config = {k: v for (k, v) in new_config.items() if v is not None}
+
             if timestamp is _SENTINEL:
                 # Keep the existing timestamp if not overridden.
                 timestamp = self._timestamp_ms
             if threadpool is _SENTINEL:
                 # Keep the existing threadpool if not overridden.
                 threadpool = self.threadpool
+
         return type(self)(
             tiledb_config=tiledb_config,
             tiledb_ctx=tiledb_ctx,
-            timestamp=timestamp,
-            threadpool=threadpool,
+            timestamp=cast(Optional[OpenTimestamp], timestamp),
+            threadpool=cast(Optional[ThreadPoolExecutor], threadpool),
```

Comment:
Surprised I had to resort to `cast`ing here. I thought mypy would understand that the `if`s on L265-L270 above should remove `SENTINEL` from the `timestamp` and `threadpool` Unions, but mypy was erroring.

Previously, the `# type: ignore[assignment]`s on L224-L225 were working around the issue. Localizing the issue here seems cleaner, but open to ideas.

Reply:
It can't infer that `_SENTINEL` is the only instance of a Type that we will accept; i.e., it thinks that there might be some other type, like `class MyBogusSentinel(_soma_tiledb_context.SENTINEL): pass`. This would mean that a user could pass `ctx.replace(timestamp=MyBogusSentinel)`, and doing so is valid within the constraints of the type system.

To do this without requiring changes to the type-checker would require the `if whatever is _SENTINEL` lines to be changed to `if isinstance(whatever, type)` (since `type` is the type of the `_SENTINEL` class-value).

I think there are a few reasonable solutions here:

 1. Change the if-statement to check whether the value in question is a type (since `_SENTINEL` is a class rather than an instance of the class):

    ```python
    def some_func(
        very_optional_param: Union[None, int, Type[_SENTINEL]] = _SENTINEL,
    ) -> ...:
      if isinstance(very_optional_param, type):
        very_optional_param = int_from_somewhere()
      # All types are gone from possible values of very_optional_param
      # leaving it as an Optional[int]
    ```

    (You can’t do `issubclass(very_optional_param, _SENTINEL)` because the first parameter to `issubclass` must be a class itself.)

 2. Use an _instance_ of a sentinel type as a default value, rather than the type itself. This might look like:

    ```python
    # this is an empty attrs class so it is immutable
    @attrs.define()
    class _Unset:
      pass

    def some_func(
        very_optional_param: Union[None, int, _Unset] = _Unset(),
    ) -> ...:
      if isinstance(very_optional_param, _Unset):
        very_optional_param = int_from_somewhere()
      # now very_optional_param is definitely an Optional[int]
      # since all possible _Sentinels are gone
      ...
    ```

 3. A simpler solution for this particular case: Make the default value a simple literal that doesn’t belong in any of the other types.

    ```python
    # My thinking behind defining these literals and constants:
    # We can document that people shouldn't be relying on
    # the exact value "__unset__" if they use the API.
    _Unset = Literal["__unset__"]
    _UNSET: _Unset = "__unset__"

    def some_func(
        very_optional_param: Union[None, int, _Unset] = _UNSET,
    ) -> ...:
      if very_optional_param == _UNSET:
        very_optional_param = int_from_somewhere()
      # "__unset__" is also now gone from the possible values
      # of very_optional_param.
    ```

I actually like the latter, since it means we just use types and instances that are already there, doesn’t have the weirdness of `Type[_SENTINEL]`, and we can always upgrade it to the immutable-class-instance–based system by changing internal code without affecting the appearance of the signature.

It would be neat if `Literal` accepted just a raw object:

```python
_SENTINEL = object()
_LiteralSentinel = Literal[_SENTINEL]
```

to mean that you could only pass the literal object `my_module._SENTINEL` the function, but it doesn’t (for sensible reasons). But if it did, we could avoid all this!

Reply:
Thanks @thetorpedodog, great explanation!

I've updated this to use `_Unset`/`_UNSET`, per your example. Typeguard and mypy seem happy with it, and it seems cleaner than my previous `cast`-based workaround.